### PR TITLE
Pull Request for Issues 11 to 15

### PR DESCRIPTION
**Pull request by syvanen**
This PR contains Jenkins Library `vars` that resolve following issues: #11, #12, #13, #14 and #15.

Will add a complete example how to use them into https://github.com/redhat-cop/container-pipelines or provide it as part of this repo.

### CODE REVIEW

File: vars/crossClusterPromote.groovy [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+		openshift.withProject("${input.sourceImagePath}") {
+			def localRegistry = openshift.selector( "is", "${input.sourceImageName}").object().status.dockerImageRepository
+			def from = "docker://${localRegistry}:${input.sourceImageTag}"
+			def to = "docker://${registry}/${input.destinationImagePath}/${input.destinationImageName}:${input.destinationImageTag}"
+
+			echo "Now Promoting ${from} -> ${to}"
+			sh """
+				set +x
+				skopeo copy --remove-signatures \
```
Comment (etsauer) [inline vars/crossClusterPromote.groovy]: @syvanen one thing to look into... the openshift team has added native functionality for moving images between regstries, which we might want to look into to replace skopeo. Take a look at `oc image mirror`.

Comment (syvanen) [reply vars/crossClusterPromote.groovy]: @etsauer That is a good point but last time I checked it didn't use the openshift authentication at all. Instead it requires Docker authentication `docker login` to function. From the help for `oc image mirror`:

> The default docker credentials are used for authenticating to the registries. 

So I'm not sure how that would work in our Jenkins slave images.

Comment (Gl4di4torRr) [reply vars/crossClusterPromote.groovy]: @syvanen adding to @etsauer comments, you may want to check out https://github.com/openshift/jenkins/issues/632 as well.

Comment (syvanen) [reply vars/crossClusterPromote.groovy]: thanks @etsauer and @Gl4di4torRr, I'll test the copying with that and update this PR to use that method.

**Comment by etsauer** [comment]
@syvanen regarding testability of new library methods, you could use https://github.com/redhat-cop/pipeline-library/tree/master/test/org/redhatcop/util as an example... basically, its a tiny pipeline deployed via an applier inventory

File: vars/applyTemplate.groovy [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env groovy
+
+class ApplyTemplateInput implements Serializable{
+  String projectName
+  String templateFile
+  String parameterFile
```
Comment (pabrahamsson) [inline vars/applyTemplate.groovy]: It seems that a default value of `""` for the `parameterFile` will let you apply a template without a parameter file. Command line test with `--param-file "" --ignore-unknown-parameters` will let me use the default value for a parameter specified in the template.

File: vars/applyTemplate.groovy [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	String clusterUrl = ""
+	String clusterToken = ""
+}
+
+def call(Map input) {
+  call(new ApplyTemplateInput(input))
+} 
+
+def call(ApplyTemplateInput input) {
+  openshift.withCluster(input.clusterUrl, input.clusterToken) {
```
Comment (pabrahamsson) [inline vars/applyTemplate.groovy]: Assertions for input parameters?

File: vars/binaryBuild.groovy [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    if(!artifactsDirectoryName?.trim()) artifactsDirectoryName = "deploy"
+    return this
+  }
+}
+
+def call (Map input){
+     call(new BinaryBuildInput(input).init())
+}
+
+def call(BinaryBuildInput input){
```
Comment (pabrahamsson) [inline vars/binaryBuild.groovy]: Assertions for input parameters?

File: vars/verifyDeployment.groovy [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+	String clusterUrl = ""
+	String clusterToken = ""
+}
+
+// verify deployment
+def call(Map input) {
+  call(new ClusterInput(input))
+}
+
+def call(ClusterInput input) {
```
Comment (pabrahamsson) [inline vars/verifyDeployment.groovy]: Assertions for input parameters?

**Comment by syvanen** [comment]
@pabrahamsson I will work on the changes once I'm back from my PTO next week.

File: vars/binaryBuild.groovy [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  openshift.withCluster() {
+    openshift.withProject("${input.project}") {
```
[the remark this answers is not in the document]
Comment (syvanen) [reply vars/binaryBuild.groovy]: True, missed it while cleaning up my code.

**Comment by syvanen** [comment]
Currently just missing test Jenkinsfile for these vars.

**Comment by pabrahamsson** [comment]
Thank you @syvanen, I'll take a look now that **I'm** back from PTO :parasol_on_ground:

File: vars/applyTemplate.groovy [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+                    
+                    // We will check if this DeploymentConfig already pre-exists and fetch the current value of Image
+                    // And set this Image -value into the DeploymentConfig template we are applying
+                    def dcSelector = openshift.selector("deploymentconfig/${o.metadata.name}")
+                    def foundObjects = dcSelector.exists()
+                    if (foundObjects) { 
+                    echo "This DC exists, copying the image value"
+                    def dcObjs = dcSelector.objects( exportable:true )
+                    echo "Image now: ${dcObjs[0].spec.template.spec.containers[0].image}"
+                    o.spec.template.spec.containers[0].image = dcObjs[0].spec.template.spec.containers[0].image
```
Comment (pabrahamsson) [inline vars/applyTemplate.groovy]: The four lines above needs to be indented one more level

File: vars/verifyDeployment.groovy [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    openshift.withCluster(input.clusterUrl, input.clusterToken) {
+        openshift.withProject("${input.projectName}") {
+            def dcObj = openshift.selector("dc", input.targetApp).object()
+            def podSelector = openshift.selector("pod", [deployment: "${input.targetApp}-${dcObj.status.latestVersion}"])
+            podSelector.untilEach {
+            echo "pod: ${it.name()}"
+            return it.object().status.containerStatuses[0].ready
```
Comment (pabrahamsson) [inline vars/verifyDeployment.groovy]: Last two lines needs one more level of indentation

**Comment by pabrahamsson** [comment]
@syvanen when I'm trying the end-to-end test on a 3.10 cluster I'm getting the following error in Jenkins, can you please take a look?
```
OpenShift Build pipeline-library-test/spring-rest-pipeline-1 from https://github.com/Gl4di4torRr/pipeline-library.git
Checking out git https://github.com/Gl4di4torRr/pipeline-library.git into /var/lib/jenkins/jobs/pipeline-library-test/jobs/pipeline-library-test-spring-rest-pipeline/workspace@script to read test/org/redhatcop/util/rollback/Jenkinsfile
Cloning the remote Git repository
Cloning repository https://github.com/Gl4di4torRr/pipeline-library.git
 > git init /var/lib/jenkins/jobs/pipeline-library-test/jobs/pipeline-library-test-spring-rest-pipeline/workspace@script # timeout=10
Fetching upstream changes from https://github.com/Gl4di4torRr/pipeline-library.git
 > git --version # timeout=10
 > git fetch --tags --progress https://github.com/Gl4di4torRr/pipeline-library.git +refs/heads/*:refs/remotes/origin/*
 > git config remote.origin.url https://github.com/Gl4di4torRr/pipeline-library.git # timeout=10
 > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git config remote.origin.url https://github.com/Gl4di4torRr/pipeline-library.git # timeout=10
Fetching upstream changes from https://github.com/Gl4di4torRr/pipeline-library.git
 > git fetch --tags --progress https://github.com/Gl4di4torRr/pipeline-library.git +refs/heads/*:refs/remotes/origin/*
 > git rev-parse origin/rollback^{commit} # timeout=10
Checking out Revision 96419ed51e2442949701f6b70fd62a1f128e587d (origin/rollback)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 96419ed51e2442949701f6b70fd62a1f128e587d
Commit message: "add rollback functionality"
First time build. Skipping changelog.
Loading library cop-library@shared-lib
Attempting to resolve shared-lib from remote references...
 > git --version # timeout=10
 > git ls-remote -h -t https://github.com/etsauer/pipeline-library.git # timeout=10
Could not find shared-lib in remote references. Pulling heads to local for deep search...
Creating git repository in /var/lib/jenkins/caches/git-9180de50c96a57452b40dcba509c2db0
 > git init /var/lib/jenkins/caches/git-9180de50c96a57452b40dcba509c2db0 # timeout=10
Setting origin to https://github.com/etsauer/pipeline-library.git
 > git config remote.origin.url https://github.com/etsauer/pipeline-library.git # timeout=10
Fetching origin...
Fetching upstream changes from origin
 > git --version # timeout=10
 > git fetch --tags --progress origin +refs/heads/*:refs/remotes/origin/*
 > git rev-parse shared-lib^{commit} # timeout=10
ERROR: Could not resolve shared-lib
hudson.plugins.git.GitException: Command "git rev-parse shared-lib^{commit}" returned status code 128:
stdout: shared-lib^{commit}
```

**Comment by syvanen** [comment]
> @syvanen when I'm trying the end-to-end test on a 3.10 cluster I'm getting the following error in Jenkins, can you please take a look?
> 
> ```
> OpenShift Build pipeline-library-test/spring-rest-pipeline-1 from https://github.com/Gl4di4torRr/pipeline-library.git
> Checking out git https://github.com/Gl4di4torRr/pipeline-library.git into /var/lib/jenkins/jobs/pipeline-library-test/jobs/pipeline-library-test-spring-rest-pipeline/workspace@script to read test/org/redhatcop/util/rollback/Jenkinsfile
> ....
> First time build. Skipping changelog.
> Loading library cop-library@shared-lib
> Attempting to resolve shared-lib from remote references...
> ....
> ERROR: Could not resolve shared-lib
> hudson.plugins.git.GitException: Command "git rev-parse shared-lib^{commit}" returned status code 128:
> stdout: shared-lib^{commit}
> ```

@pabrahamsson I think you might have referenced to wrong branch or tag of your Git repo.

What did you have as library identifier? For example here's mine:
> library identifier: "pipeline-library@issues-11-to-15", retriever: modernSCM(
>   [$class: "GitSCMSource",
>    remote: "https://github.com/syvanen/pipeline-library.git"])

Where "pipeline-library" is the name for library and "@issues-11-to-15" is the git branch/tag name used in my own personal fork.

File: vars/applyTemplate.txt [hunk 9]
```diff
@@ -0,0 +1,30 @@
+# applyTemplate
+
+This function processes template file with parameters file and then runs "oc apply" for the processing results.
+Also it has a workaround for the bug that if the DeploymentConfig already exists, it will break the Image Trigger.
```
[the remark this answers is not in the document]
Comment (syvanen) [reply vars/applyTemplate.txt]: I'll have to check this later this week, how to reproduce the error and then see results.

**Comment by syvanen** [comment]
@etsauer and @pabrahamsson now the Jenkins file covers `binaryBuild`, `verifyDeployment` and `tagImage`.

Would this be ok to merge? I will work on adding the `applyTemplate` and `crossClusterPromote` also into the pipeline. But that will require a bit more work.

**Pull request merged**